File: src/lib/core/FrameIndexGenerator.cpp
/*************************************************************************
 * Dicomifier - Copyright (C) Universite de Strasbourg
 * Distributed under the terms of the CeCILL-B license, as published by
 * the CEA-CNRS-INRIA. Refer to the LICENSE file or to
 * http://www.cecill.info/licences/Licence_CeCILL-B_V1-en.html
 * for details.
 ************************************************************************/
#include <algorithm>
#include "FrameIndexGenerator.h"

namespace dicomifier
{
    
FrameIndexGenerator
::FrameIndexGenerator(std::vector<int> indexmax, int countmax)
    :_indexMax(indexmax), _currentIndex({}), _countMax(countmax), _currentStep(0)
{
    if (!indexmax.empty())
    {
        this->_currentIndex = std::vector<int>(indexmax.size(), 0);
    }
    
    if (this->_countMax == -1)
    {
        this->computeCountMax();
    }
}

FrameIndexGenerator
::~FrameIndexGenerator()
{
    // Nothing to do
}

bool
FrameIndexGenerator
::done() const
{
    return (this->_currentStep >= this->_countMax);
}

void
FrameIndexGenerator
::next()
{
    _currentStep++;
    
    for (auto index = this->_currentIndex.size()-1; index != -1; index--)
    {
        // Increase current index
        this->_currentIndex[index]++;
        
        // If index reaches the max
        if (this->_currentIndex[index] >= this->_indexMax[index])
        {
            // Restart to 0
            this->_currentIndex[index] = 0;
            // continue, go to the next index
        }
        else
        {
            // increase done => stop
            break;
        }
    }
}

std::vector<int> 
FrameIndexGenerator
::get_index() const
{
    return this->_currentIndex;
}

std::vector<int> 
FrameIndexGenerator
::get_indexMax() const
{
    return this->_indexMax;
}

int 
FrameIndexGenerator
::get_step() const
{
    return this->_currentStep;
}

int 
FrameIndexGenerator
::get_countMax() const
{
    return this->_countMax;
}

void 
FrameIndexGenerator
::computeCountMax()
{
    int max = 1;
    for (auto value : this->_indexMax)
    {
        if (value != 0)
        {
            max *= value;
        }
    }
    this->_countMax = max;
}

int 
FrameIndexGenerator
::compute_index(std::vector<int> withoutindex) const
{
    std::vector<int> vect;
    for (auto i = 0; i < this->_currentIndex.size(); i++)
    {
        if (std::find(withoutindex.begin(), 
                      withoutindex.end(), i) == withoutindex.end())
        {
            for (auto j = 0; j < vect.size(); j++)
            {
                vect[j] = vect[j] * this->_indexMax[i];
            }
            vect.push_back(this->_currentIndex[i]);
        }
    }
    
    int ret = 0;
    for (auto value : vect)
    {
        ret += value;
    }
    return ret;
}
    
} // namespace dicomifier

File: src/python/dicomifier/bruker_to_dicom/frame_index_generator.py
#########################################################################
# Dicomifier - Copyright (C) Universite de Strasbourg
# Distributed under the terms of the CeCILL-B license, as published by
# the CEA-CNRS-INRIA. Refer to the LICENSE file or to
# http://www.cecill.info/licences/Licence_CeCILL-B_V1-en.html
# for details.
#########################################################################

import itertools

import numpy

class FrameIndexGenerator(object):
    """ Generate the indices to iterate through the frame groups of a Bruker
        data set.
    """
    
    def __init__(self, data_set):
        self._frame_groups = []
        self._frames_count = 0
        self._dependent_fields = []
        self._strides = []
        
        frame_groups = []
        for description in data_set.get("VisuFGOrderDesc", []):
            frame_count, name, _, begin, fields_count = description
            if fields_count == 0:
                # This may happen with Paravision 5
                fields = []
            else:
                fields = [
                    x[0] for x in 
                    data_set["VisuGroupDepVals"][begin:begin+fields_count]]
            frame_groups.append([frame_count, name, fields])
        # WARNING: the frame groups are listed in innermost-to-outermost
        # order, while FrameIndexGenerator uses outermost-to-innermost order.
        # Invert now, to match the order of FrameIndexGenerator.
        self.frame_groups = frame_groups[::-1]
    
    def __iter__(self):
        return itertools.product(
            *[range(count) for count, _, _ in self.frame_groups])
    
    def get_linear_index(self, index):
        """ Return the linear index associated to the multi-dimensional index.
        """
        
        return numpy.dot(index[::-1], self._strides) if self._frames_count else 0
    
    @property
    def frame_groups(self):
        return self._frame_groups
    
    @frame_groups.setter
    def frame_groups(self, value):
        self._frame_groups = value
        
        shape = [count for count, _, _ in self._frame_groups][::-1]
        
        self._frames_count = numpy.cumprod(shape)[-1] if shape else 0
        self._dependent_fields = list(
            itertools.chain(*[fields for _, _, fields in self.frame_groups]))
        self._strides = numpy.cumprod([1]+shape[:-1])
    
    @property
    def frames_count(self):
        """ Total number of frames.
        """
        
        return max(1, self._frames_count)
    
    @property
    def dependent_fields(self):
        """ Names of all fields mentioned in the frame groups.
        """
        
        return self._dependent_fields
